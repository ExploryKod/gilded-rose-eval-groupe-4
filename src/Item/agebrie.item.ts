import { changeSellIn } from "../SellInDelay";
import { changeQuality } from "../QualityStatus";

export class AgedBrie extends changeQuality {
  constructor(name, sellIn, quality) {
    super(name, sellIn, quality);
  }

  changeSellIn = new changeSellIn('Age Brie', this.sellIn, this.quality)

  update() {
    this.increaseQuality();
    this.changeSellIn.decreaseSellIn(1);
    if (this.sellIn < 0) {
      console.log(`Délai de vente dépassé`);
      console.log(`La qualité de ${this.name} augmente quand même`);
      this.increaseQuality();
    }
  }

}