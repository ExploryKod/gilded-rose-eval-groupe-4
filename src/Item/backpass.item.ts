import { changeSellIn } from "../SellInDelay";
import { changeQuality } from "../QualityStatus";

export class BackstagePasses extends changeQuality {

  constructor(name, sellIn, quality) {
    super(name, sellIn, quality);
  }

  changeSellIn = new changeSellIn('Backstage passes to a TAFKAL80ETC concert', this.sellIn, this.quality)

  public update() {

    if (this.quality < 50) {

      this.increaseQuality();

      console.log(`SellIn est égal à ${this.sellIn}`);

      const bonusThresholds = [11, 6];

      bonusThresholds.forEach(threshold => {
        if (this.sellIn < threshold) {
          console.log(`SellIn est inférieur à ${threshold}`);
          this.increaseQuality();
        }
      });
    }

    this.changeSellIn.decreaseSellIn(1);

    if (this.sellIn < 0) {
      console.log(`Délai de vente dépassé`);
      this.setQualityToZero();
    }
  }
}