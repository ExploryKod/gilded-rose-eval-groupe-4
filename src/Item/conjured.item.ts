import { changeSellIn } from "../SellInDelay";
import { changeQuality } from "../QualityStatus";

export class Conjured extends changeQuality {
  constructor(name, sellIn, quality) {
    super(name, sellIn, quality);
  }

  changeSellIn = new changeSellIn('Conjured', this.sellIn, this.quality)

  update() {
    this.decreaseQuality(2);
    this.changeSellIn.decreaseSellIn(1);
  }
}
