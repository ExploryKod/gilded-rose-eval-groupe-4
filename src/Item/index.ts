type ItemsName =
  "Aged Brie"
  | "Backstage passes to a TAFKAL80ETC concert"
  | "Sulfuras, Hand of Ragnaros"
  | "Conjured";

export class Item {
  name: ItemsName;
  sellIn: number;
  quality: number;
  BonusThresholds: Array<number>;

  constructor(name, sellIn, quality, BonusThresholds = []) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
    this.BonusThresholds = BonusThresholds;
  }
}

export class Items {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }
}



