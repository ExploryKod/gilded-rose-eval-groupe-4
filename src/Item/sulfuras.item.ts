import { Item } from "../Item";
import { changeSellIn } from "../SellInDelay";
import { changeQuality } from "../QualityStatus";

export class Sulfuras extends Item {
  constructor(name, sellIn, quality) {
    super(name, sellIn, quality);
  }

  update(): void {
  }
}