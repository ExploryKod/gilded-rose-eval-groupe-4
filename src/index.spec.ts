import runGoldenMaster from "jest-golden-master";
import {GildedRose } from "./index";
import {Item} from "./Item";

test("1 test with Sulfuras item", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Sulfuras, Hand of Ragnaros", 0, 80);
    const items = new GildedRose([item1]);
    items.updateQuality();
  });
});

test("2 test with Aged Brie updated 1x with 10 for sellIn and quality", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Aged Brie", 10, 10 );
    const items = new GildedRose([item1]);
    items.updateQuality();
  });
});

test("3 test with Aged Brie when sellIn is negative", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Aged Brie", 2, 10 );
    const items = new GildedRose([item1]);
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
  });
});

test("4 test with Backstage passes to a TAFKAL80ETC concert", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Backstage passes to a TAFKAL80ETC concert", 5, 10 );
    const items = new GildedRose([item1]);
    items.updateQuality();
    items.updateQuality();
  });
});

test("5 test with Backstage passes to a TAFKAL80ETC concert when sellIn is negative", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Backstage passes to a TAFKAL80ETC concert", 2, 10 );
    const items = new GildedRose([item1]);
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
  });
});

test("6 test with several items when quality goes over 50", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Backstage passes to a TAFKAL80ETC concert", 15, 48 );
    const item2 = new Item("Aged Brie", 8, 48 );
    const items = new GildedRose([item1, item2]);
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
  });
});

test("7 test quality under 50", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Backstage passes to a TAFKAL80ETC concert", 15, -4 );
    const item2 = new Item("Aged Brie", 8, 23 );
    const items = new GildedRose([item1, item2]);
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
  });
});

test("8 test with Conjured item", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Conjured", 15, 3 );
    const items = new GildedRose([item1]);
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
  });
});

test("9 test with Conjured item when sellIn goes negative", async () => {
  runGoldenMaster(async () => {
    const item1 = new Item("Conjured", 2, 47 );
    const items = new GildedRose([item1]);
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
    items.updateQuality();
  });
});


/////////////////////// SIMULATION OF CONJURED BEFORE IT EXISTS //////////////////////
// test("99 test without category normally", async () => {
//   runGoldenMaster(async () => {
//     const item1 = new Item("normal item", 15, 3 );
//     const items = new GildedRose([item1]);
//     items.updateQuality();
//     items.updateQuality();
//     items.updateQuality();
//     items.updateQuality();
//   });
// });
//
// test("99 test without category when sellIn goes negative", async () => {
//   runGoldenMaster(async () => {
//     const item1 = new Item("no category", 3, 45 );
//     const items = new GildedRose([item1]);
//     items.updateQuality();
//     items.updateQuality();
//     items.updateQuality();
//     items.updateQuality();
//   });
// });