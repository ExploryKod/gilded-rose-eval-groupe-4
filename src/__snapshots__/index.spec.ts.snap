// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 test with Sulfuras item 1`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Sulfuras, Hand of Ragnaros avec une qualité de 80",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;

exports[`2 test with Aged Brie updated 1x with 10 for sellIn and quality 1`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Aged Brie avec une qualité de 10",
  "La qualité augmente de 1, elle est maintenant de 11",
  "SellIn prend -1 et est maintenant de 9",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;

exports[`3 test with Aged Brie when sellIn is negative 1`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Aged Brie avec une qualité de 10",
  "La qualité augmente de 1, elle est maintenant de 11",
  "SellIn prend -1 et est maintenant de 1",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Aged Brie avec une qualité de 11",
  "La qualité augmente de 1, elle est maintenant de 12",
  "SellIn prend -1 et est maintenant de 0",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Aged Brie avec une qualité de 12",
  "La qualité augmente de 1, elle est maintenant de 13",
  "SellIn prend -1 et est maintenant de -1",
  "Délai de vente dépassé",
  "La qualité de Aged Brie augmente quand même",
  "La qualité augmente de 1, elle est maintenant de 14",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;

exports[`4 test with Backstage passes to a TAFKAL80ETC concert when 1`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 10",
  "La qualité augmente de 1, elle est maintenant de 11",
  "SellIn est égal à 5",
  "SellIn est inférieur à 11",
  "La qualité augmente de 1, elle est maintenant de 12",
  "SellIn est inférieur à 6",
  "La qualité augmente de 1, elle est maintenant de 13",
  "SellIn prend -1 et est maintenant de 4",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 13",
  "La qualité augmente de 1, elle est maintenant de 14",
  "SellIn est égal à 4",
  "SellIn est inférieur à 11",
  "La qualité augmente de 1, elle est maintenant de 15",
  "SellIn est inférieur à 6",
  "La qualité augmente de 1, elle est maintenant de 16",
  "SellIn prend -1 et est maintenant de 3",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;

exports[`5 test with Backstage passes to a TAFKAL80ETC concert when sellIn is negative 1`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 10",
  "La qualité augmente de 1, elle est maintenant de 11",
  "SellIn est égal à 2",
  "SellIn est inférieur à 11",
  "La qualité augmente de 1, elle est maintenant de 12",
  "SellIn est inférieur à 6",
  "La qualité augmente de 1, elle est maintenant de 13",
  "SellIn prend -1 et est maintenant de 1",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 13",
  "La qualité augmente de 1, elle est maintenant de 14",
  "SellIn est égal à 1",
  "SellIn est inférieur à 11",
  "La qualité augmente de 1, elle est maintenant de 15",
  "SellIn est inférieur à 6",
  "La qualité augmente de 1, elle est maintenant de 16",
  "SellIn prend -1 et est maintenant de 0",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 16",
  "La qualité augmente de 1, elle est maintenant de 17",
  "SellIn est égal à 0",
  "SellIn est inférieur à 11",
  "La qualité augmente de 1, elle est maintenant de 18",
  "SellIn est inférieur à 6",
  "La qualité augmente de 1, elle est maintenant de 19",
  "SellIn prend -1 et est maintenant de -1",
  "Délai de vente dépassé",
  "Délai de vente dépassé : la qualité de Backstage passes to a TAFKAL80ETC concert est maintenant de 0",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;

exports[`6 test with several items when quality goes over 50 1`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 48",
  "La qualité augmente de 1, elle est maintenant de 49",
  "SellIn est égal à 15",
  "SellIn prend -1 et est maintenant de 14",
  "Item n° 2 - nom : Aged Brie avec une qualité de 48",
  "La qualité augmente de 1, elle est maintenant de 49",
  "SellIn prend -1 et est maintenant de 7",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 49",
  "La qualité augmente de 1, elle est maintenant de 50",
  "SellIn est égal à 14",
  "SellIn prend -1 et est maintenant de 13",
  "Item n° 2 - nom : Aged Brie avec une qualité de 49",
  "La qualité augmente de 1, elle est maintenant de 50",
  "SellIn prend -1 et est maintenant de 6",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 50",
  "SellIn prend -1 et est maintenant de 12",
  "Item n° 2 - nom : Aged Brie avec une qualité de 50",
  "SellIn prend -1 et est maintenant de 5",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de 50",
  "SellIn prend -1 et est maintenant de 11",
  "Item n° 2 - nom : Aged Brie avec une qualité de 50",
  "SellIn prend -1 et est maintenant de 4",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;

exports[`7 test quality under 50 1`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de -4",
  "La qualité augmente de 1, elle est maintenant de -3",
  "SellIn est égal à 15",
  "SellIn prend -1 et est maintenant de 14",
  "Item n° 2 - nom : Aged Brie avec une qualité de 23",
  "La qualité augmente de 1, elle est maintenant de 24",
  "SellIn prend -1 et est maintenant de 7",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de -3",
  "La qualité augmente de 1, elle est maintenant de -2",
  "SellIn est égal à 14",
  "SellIn prend -1 et est maintenant de 13",
  "Item n° 2 - nom : Aged Brie avec une qualité de 24",
  "La qualité augmente de 1, elle est maintenant de 25",
  "SellIn prend -1 et est maintenant de 6",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de -2",
  "La qualité augmente de 1, elle est maintenant de -1",
  "SellIn est égal à 13",
  "SellIn prend -1 et est maintenant de 12",
  "Item n° 2 - nom : Aged Brie avec une qualité de 25",
  "La qualité augmente de 1, elle est maintenant de 26",
  "SellIn prend -1 et est maintenant de 5",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Backstage passes to a TAFKAL80ETC concert avec une qualité de -1",
  "La qualité augmente de 1, elle est maintenant de 0",
  "SellIn est égal à 12",
  "SellIn prend -1 et est maintenant de 11",
  "Item n° 2 - nom : Aged Brie avec une qualité de 26",
  "La qualité augmente de 1, elle est maintenant de 27",
  "SellIn prend -1 et est maintenant de 4",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;

exports[`99 test with conjured 1`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Conjured avec une qualité de 3",
  "La qualité diminue de 1, elle est maintenant de 2",
  "SellIn prend -1 et est maintenant de 14",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Conjured avec une qualité de 2",
  "La qualité diminue de 1, elle est maintenant de 1",
  "SellIn prend -1 et est maintenant de 13",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Conjured avec une qualité de 1",
  "La qualité diminue de 1, elle est maintenant de 0",
  "SellIn prend -1 et est maintenant de 12",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Conjured avec une qualité de 0",
  "SellIn prend -1 et est maintenant de 11",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;

exports[`99 test with conjured 2`] = `
[
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Conjured avec une qualité de 45",
  "La qualité diminue de 1, elle est maintenant de 44",
  "SellIn prend -1 et est maintenant de 2",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Conjured avec une qualité de 44",
  "La qualité diminue de 1, elle est maintenant de 43",
  "SellIn prend -1 et est maintenant de 1",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Conjured avec une qualité de 43",
  "La qualité diminue de 1, elle est maintenant de 42",
  "SellIn prend -1 et est maintenant de 0",
  "Fin de la mise à jour quotidienne de la qualité des items",
  "Mise à jour quotidienne de la qualité des items",
  "Item n° 1 - nom : Conjured avec une qualité de 42",
  "La qualité diminue de 1, elle est maintenant de 41",
  "SellIn prend -1 et est maintenant de -1",
  "Délai de vente dépassé",
  "La qualité diminue de 1, elle est maintenant de 40",
  "Fin de la mise à jour quotidienne de la qualité des items",
]
`;
